crypto/cmp/cmp_client.c

```diff
@@ -0,0 +1,878 @@
+                                CMP_R_POLLREP_NOT_RECEIVED))
+            goto err;
+
+        /* handle potential pollRep */
+        if (ossl_cmp_msg_get_bodytype(prep) == OSSL_CMP_PKIBODY_POLLREP) {
+            int64_t check_after;
+            OSSL_CMP_POLLREPCONTENT *prc = prep->body->value.pollRep;
+
+            /*
+             * TODO: handle multiple elements, in case multiple requests have
+             * been sent - see https://github.com/mpeylo/cmpossl/issues/67
+             */
+            if (sk_OSSL_CMP_POLLREP_num(prc) > 1) {
+                CMPerr(0, CMP_R_MULTIPLE_RESPONSES_NOT_SUPPORTED);
+                goto err;
+            }
+            if ((pollRep = ossl_cmp_pollrepcontent_get0_pollrep(prc, rid))
+                == NULL)
+                goto err;
+            if (!ASN1_INTEGER_get_int64(&check_after, pollRep->checkAfter)) {
+                CMPerr(0, CMP_R_BAD_CHECKAFTER_IN_POLLREP);
+                goto err;
+            }
+            if (check_after < 0) {
+                CMPerr(0, CMP_R_RECEIVED_NEGATIVE_CHECKAFTER_IN_POLLREP);
+                goto err;
+            }
+            /* TODO: print OPTIONAL reason (PKIFreeText) from message */
+            ossl_cmp_log1(INFO, ctx,
+                          "received polling response, waiting for check_after = %ld seconds",
+                          check_after);
+
+            if (ctx->totaltimeout > 0) { /* total timeout is not infinite */
+                const int exp = 5; /* expected max time per msg round trip */
+                int64_t time_left = (int64_t)(ctx->end_time - exp - time(NULL));
+
+                if (time_left <= 0) {
+                    CMPerr(0, CMP_R_TOTAL_TIMEOUT);
+                    goto err;
+                }
+                if (time_left < check_after)
+                    check_after = time_left;
+                /* poll one last time just when timeout was reached */
+            }
+
+            OSSL_CMP_MSG_free(preq);
+            preq = NULL;
+            OSSL_CMP_MSG_free(prep);
+            prep = NULL;
+            sleep((unsigned int)check_after);
+        } else {
+            ossl_cmp_info(ctx, "got ip/cp/kup after polling");
+            break;
+        }
+    }
+    if (prep == NULL)
+        goto err;
+
+    OSSL_CMP_MSG_free(preq);
+    *out = prep;
+
+    return 1;
+ err:
+    OSSL_CMP_MSG_free(preq);
+    OSSL_CMP_MSG_free(prep);
+    return 0;
+}
+
+/* Send certConf for IR, CR or KUR sequences and check response */
+int ossl_cmp_exchange_certConf(OSSL_CMP_CTX *ctx, int fail_info,
+                               const char *txt)
+{
+    OSSL_CMP_MSG *certConf = NULL;
+    OSSL_CMP_MSG *PKIconf = NULL;
+    int res = 0;
+
+    /* OSSL_CMP_certConf_new() also checks if all necessary options are set */
+    if ((certConf = ossl_cmp_certConf_new(ctx, fail_info, txt)) == NULL)
+        goto err;
+
+    res = send_receive_check(ctx, certConf, &PKIconf, OSSL_CMP_PKIBODY_PKICONF,
```

Comment:
It seems we throw away the returned value in the third argument in a few of these calls. I wonder whether it is worth allowing NULL to be provided there.

Reply:
The response message is thrown away only in case of a `PKIConf` response, which has an empty body and is basically just an ACK.
I've checked: allowing a NULL pointer result variable would just complicate matters because then some case distinctions in `send_receive_check()` would be needed and in any case the response needs to be deallocated.

Reply:
Ok